File: mobile/lib/services/booking_intent_service.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import '../config/api_config.dart';
import '../models/booking_intent.dart';
import 'storage_service.dart';

/// Service để xử lý BookingIntent - Temporary Reservation/Locking mechanism
/// Prevents overbooking when multiple users try to book the same listing
class BookingIntentService {
  final StorageService _storageService = StorageService();

  /// Check if a listing is available for the specified dates
  Future<Map<String, dynamic>> checkAvailability({
    required String listingId,
    required DateTime startDate,
    required DateTime endDate,
    required String userId,
  }) async {
    try {
      final token = await _storageService.getToken();

      final queryParams = {
        'startDate': startDate.toIso8601String(),
        'endDate': endDate.toIso8601String(),
        'userId': userId,
      };

      final uri = Uri.parse(
        '${ApiConfig.baseUrl}/booking-intent/check-availability/$listingId'
      ).replace(queryParameters: queryParams);

      debugPrint('🔍 Checking availability: $uri');

      final response = await http.get(
        uri,
        headers: ApiConfig.headers(token: token),
      );

      debugPrint('📥 Availability response: ${response.statusCode}');

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return {
          'success': true,
          'available': data['available'] ?? false,
          'lockedBy': data['lockedBy'],
          'lockedUntil': data['lockedUntil'],
          'message': data['message'],
        };
      } else {
        final error = json.decode(response.body);
        return {
          'success': false,
          'available': false,
          'message': error['message'] ?? 'Failed to check availability',
        };
      }
    } catch (e) {
      debugPrint('❌ Error checking availability: $e');
      return {
        'success': false,
        'available': false,
        'message': 'An error occurred: ${e.toString()}',
      };
    }
  }

  /// Create a booking intent (lock the listing temporarily)
  Future<Map<String, dynamic>> createBookingIntent({
    required String customerId,
    required String hostId,
    required String listingId,
    required DateTime startDate,
    required DateTime endDate,
    required double totalPrice,
    required String paymentMethod, // vnpay, cash
    required String paymentType, // full, deposit, cash
    double? depositAmount,
    int? depositPercentage,
    String bookingType = 'entire_place',
  }) async {
    try {
      final token = await _storageService.getToken();

      if (token == null) {
        return {'success': false, 'message': 'Not authenticated'};
      }

      final uri = Uri.parse('${ApiConfig.baseUrl}/booking-intent/create');

      debugPrint('🔒 Creating booking intent for listing: $listingId');
      debugPrint('📅 Dates: $startDate to $endDate');
      debugPrint('💳 Payment: $paymentMethod / $paymentType');

      // Calculate payment amount
      double paymentAmount = totalPrice;
      double remainingAmount = 0;

      if (paymentType == 'deposit') {
        final percentage = depositPercentage ?? 30;
        paymentAmount = depositAmount ?? (totalPrice * percentage / 100);
        remainingAmount = totalPrice - paymentAmount;
      }

      final body = {
        'customerId': customerId,
        'hostId': hostId,
        'listingId': listingId,
        'bookingType': bookingType,
        'startDate': startDate.toIso8601String(),
        'endDate': endDate.toIso8601String(),
        'totalPrice': totalPrice,
        'paymentMethod': paymentMethod,
        'paymentType': paymentType,
        'paymentAmount': paymentAmount,
        'depositPercentage': depositPercentage ?? 0,
        'depositAmount': depositAmount ?? 0,
        'remainingAmount': remainingAmount,
      };

      final response = await http.post(
        uri,
        headers: ApiConfig.headers(token: token),
        body: json.encode(body),
      );

      debugPrint('📥 Create intent response: ${response.statusCode}');

      if (response.statusCode == 200 || response.statusCode == 201) {
        final data = json.decode(response.body);
        debugPrint('✅ Booking intent created: ${data['intent']?['intentId']}');

        return {
          'success': true,
          'message': data['message'] ?? 'Listing locked successfully',
          'intent': data['intent'] != null
            ? BookingIntent.fromJson(data['intent'])
            : null,
          'intentId': data['intent']?['intentId'],
          'expiresAt': data['intent']?['expiresAt'],
        };
      } else if (response.statusCode == 409) {
        // Listing is locked by another user
        final error = json.decode(response.body);
        debugPrint('⚠️ Listing is locked: ${error['message']}');
        return {
          'success': false,
          'locked': true,
          'message': error['message'] ?? 'This listing is currently being booked by another user',
          'lockedUntil': error['lockedUntil'],
        };
      } else {
        final error = json.decode(response.body);
        debugPrint('❌ Failed to create intent: ${error['message']}');
        return {
          'success': false,
          'message': error['message'] ?? 'Failed to create booking intent',
        };
      }
    } catch (e) {
      debugPrint('❌ Error creating booking intent: $e');
      return {
        'success': false,
        'message': 'An error occurred: ${e.toString()}',
      };
    }
  }

  /// Confirm payment and create actual booking
  Future<Map<String, dynamic>> confirmPayment({
    required String intentId,
    required String transactionId,
    String? vnpTransactionNo,
  }) async {
    try {
      final token = await _storageService.getToken();

      if (token == null) {
        return {'success': false, 'message': 'Not authenticated'};
      }

      final uri = Uri.parse('${ApiConfig.baseUrl}/booking-intent/$intentId/confirm');

      debugPrint('💳 Confirming payment for intent: $intentId');
      debugPrint('🔖 Transaction ID: $transactionId');

      final response = await http.post(
        uri,
        headers: ApiConfig.headers(token: token),
        body: json.encode({
          'transactionId': transactionId,
          'vnpTransactionNo': vnpTransactionNo,
        }),
      );

      debugPrint('📥 Confirm payment response: ${response.statusCode}');

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        debugPrint('✅ Payment confirmed, booking created: ${data['booking']?['_id']}');
        return {
          'success': true,
          'message': data['message'] ?? 'Booking confirmed successfully',
          'booking': data['booking'],
          'bookingId': data['booking']?['_id'],
        };
      } else {
        final error = json.decode(response.body);
        debugPrint('❌ Payment confirmation failed: ${error['message']}');
        return {
          'success': false,
          'message': error['message'] ?? 'Failed to confirm payment',
        };
      }
    } catch (e) {
      debugPrint('❌ Error confirming payment: $e');
      return {
        'success': false,
        'message': 'An error occurred: ${e.toString()}',
      };
    }
  }

  /// Cancel a booking intent (release the lock)
  Future<Map<String, dynamic>> cancelIntent(String intentId) async {
    try {
      final token = await _storageService.getToken();

      if (token == null) {
        return {'success': false, 'message': 'Not authenticated'};
      }

      final uri = Uri.parse('${ApiConfig.baseUrl}/booking-intent/$intentId/cancel');

      debugPrint('🔓 Cancelling booking intent: $intentId');

      final response = await http.post(
        uri,
        headers: ApiConfig.headers(token: token),
      );

      debugPrint('📥 Cancel intent response: ${response.statusCode}');

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        debugPrint('✅ Booking intent cancelled');
        return {
          'success': true,
          'message': data['message'] ?? 'Booking intent cancelled',
        };
      } else {
        final error = json.decode(response.body);
        return {
          'success': false,
          'message': error['message'] ?? 'Failed to cancel booking intent',
        };
      }
    } catch (e) {
      debugPrint('❌ Error cancelling intent: $e');
      return {
        'success': false,
        'message': 'An error occurred: ${e.toString()}',
      };
    }
  }

  /// Get booking intent by ID
  Future<BookingIntent?> getIntent(String intentId) async {
    try {
      final token = await _storageService.getToken();

      final uri = Uri.parse('${ApiConfig.baseUrl}/booking-intent/$intentId');

      final response = await http.get(
        uri,
        headers: ApiConfig.headers(token: token),
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return BookingIntent.fromJson(data['intent'] ?? data);
      }
      return null;
    } catch (e) {
      debugPrint('❌ Error fetching intent: $e');
      return null;
    }
  }

  /// Get user's active intents
  Future<List<BookingIntent>> getUserActiveIntents(String userId) async {
    try {
      final token = await _storageService.getToken();

      final uri = Uri.parse('${ApiConfig.baseUrl}/booking-intent/user/$userId/active');

      final response = await http.get(
        uri,
        headers: ApiConfig.headers(token: token),
      );

      if (response.statusCode == 200) {
        final List<dynamic> data = json.decode(response.body);
        return data.map((json) => BookingIntent.fromJson(json)).toList();
      }
      return [];
    } catch (e) {
      debugPrint('❌ Error fetching user intents: $e');
      return [];
    }
  }
}

File: server/services/bookingIntent.service.js
const BookingIntent = require("../models/BookingIntent");
const Listing = require("../models/Listing");
const Booking = require("../models/Booking");
const {v4: uuidv4} = require("uuid");
const {BOOKING_INTENT_TIMEOUT_MINUTES} = require("../config/bookingIntentConfig");

const LOCK_TIMEOUT_MINUTES = BOOKING_INTENT_TIMEOUT_MINUTES;

const releaseExpiredLocks = async () => {
    return await BookingIntent.releaseExpiredLocks();
};

const createBookingIntent = async (data) => {
    const { customerId, hostId, listingId, bookingType, startDate, endDate, totalPrice } = data;
    const { paymentMethod, paymentType, paymentAmount, depositPercentage, depositAmount, remainingAmount } = data;

    const listing = await Listing.findById(listingId);
    if (!listing) {
        const error = new Error("Listing not found");
        error.statusCode = 404;
        throw error;
    }

    const existingBooking = await Booking.findOne({
        listingId: listingId,
        bookingStatus: {$in: ["pending", "approved", "checked_in"]},
        $or: [
            {startDate: {$lte: endDate}, endDate: {$gte: startDate}}
        ]
    });

    if (existingBooking) {
        const error = new Error("This listing is already booked for the selected dates");
        error.statusCode = 409;
        error.conflictType = "existing_booking";
        throw error;
    }

    const existingLock = await BookingIntent.getActiveLock(listingId, startDate, endDate);

    if (existingLock && existingLock.customerId.toString() !== customerId) {
        const remainingSeconds = Math.ceil((existingLock.expiresAt - new Date()) / 1000);
        const error = new Error("This listing is currently being reserved by another user");
        error.statusCode = 409;
        error.conflictType = "locked_by_other";
        error.retryAfterSeconds = remainingSeconds;
        throw error;
    }

    if (existingLock && existingLock.customerId.toString() === customerId) {
        return { bookingIntent: existingLock, isExisting: true };
    }

    const intentId = `INTENT_${Date.now()}_${uuidv4().slice(0, 8)}`;
    const tempOrderId = `TEMP_${Date.now()}_${uuidv4().slice(0, 10)}`;
    const expiresAt = new Date(Date.now() + LOCK_TIMEOUT_MINUTES * 60 * 1000);

    const bookingIntent = new BookingIntent({
        intentId,
        tempOrderId,
        customerId,
        hostId,
        listingId,
        bookingType: bookingType || "entire_place",
        startDate,
        endDate,
        totalPrice,
        status: "locked",
        paymentMethod: paymentMethod || "vnpay",
        paymentType: paymentType || "full",
        paymentAmount: paymentAmount || totalPrice,
        depositPercentage: depositPercentage || 0,
        depositAmount: depositAmount || 0,
        remainingAmount: remainingAmount || 0,
        lockedAt: new Date(),
        expiresAt,
    });

    await bookingIntent.save();
    return { bookingIntent, isExisting: false, expiresIn: LOCK_TIMEOUT_MINUTES * 60 };
};

const checkAvailability = async (listingId, startDate, endDate, userId) => {
    const existingBooking = await Booking.findOne({
        listingId: listingId,
        bookingStatus: {$in: ["pending", "approved", "checked_in"]},
        $or: [
            {startDate: {$lte: endDate}, endDate: {$gte: startDate}}
        ]
    });

    if (existingBooking) {
        return { available: false, reason: "already_booked", message: "This listing is already booked for the selected dates" };
    }

    const activeLock = await BookingIntent.getActiveLock(listingId, startDate, endDate);

    if (activeLock) {
        const isOwnLock = userId && activeLock.customerId.toString() === userId;
        const remainingSeconds = Math.ceil((activeLock.expiresAt - new Date()) / 1000);

        return {
            available: isOwnLock, // Available if it's user's own lock
            reason: isOwnLock ? "own_lock" : "locked_by_other",
            message: isOwnLock
                ? "You have an active reservation for this listing"
                : "This listing is currently being reserved by another user",
            retryAfterSeconds: isOwnLock ? null : remainingSeconds,
            bookingIntent: isOwnLock ? activeLock : null,
        };
    }

    return { available: true, message: "Listing is available for booking" };
};

const getByTempOrderId = async (tempOrderId) => {
    const bookingIntent = await BookingIntent.findOne({tempOrderId})
        .populate("customerId", "firstName lastName email")
        .populate("listingId", "title");

    if (!bookingIntent) {
        const error = new Error("Booking intent not found");
        error.statusCode = 404;
        throw error;
    }
    return bookingIntent;
};

const getById = async (intentId) => {
    const bookingIntent = await BookingIntent.findOne({intentId})
            .populate("customerId", "firstName lastName email profileImagePath")
            .populate("hostId", "firstName lastName email profileImagePath")
            .populate("listingId", "title listingPhotoPaths price");

    if (!bookingIntent) {
        const error = new Error("Booking intent not found");
        error.statusCode = 404;
        throw error;
    }

    if (bookingIntent.status === "locked" && bookingIntent.expiresAt < new Date()) {
        bookingIntent.status = "expired";
        bookingIntent.failureReason = "Lock expired";
        await bookingIntent.save();
    }
    return bookingIntent;
};

const cancelIntent = async (intentId, userId, reason) => {
    const bookingIntent = await BookingIntent.findOne({intentId});

    if (!bookingIntent) {
        const error = new Error("Booking intent not found");
        error.statusCode = 404;
        throw error;
    }

    if (bookingIntent.customerId.toString() !== userId) {
        const error = new Error("You are not authorized to cancel this booking intent");
        error.statusCode = 403;
        throw error;
    }

    if (bookingIntent.status !== "locked") {
        const error = new Error(`Cannot cancel booking intent with status: ${bookingIntent.status}`);
        error.statusCode = 400;
        throw error;
    }

    await bookingIntent.cancel(reason || "User cancelled");
    return bookingIntent;
};

const confirmIntent = async (intentId, transactionId) => {
    const bookingIntent = await BookingIntent.findOne({intentId});

    if (!bookingIntent) {
        const error = new Error("Booking intent not found");
        error.statusCode = 404;
        throw error;
    }

    if (bookingIntent.status !== "locked") {
        const error = new Error(`Cannot confirm booking intent with status: ${bookingIntent.status}`);
        error.statusCode = 400;
        throw error;
    }

    if (bookingIntent.expiresAt < new Date()) {
        await bookingIntent.fail("Lock expired before payment confirmation");
        const error = new Error("Booking lock has expired. Please start the booking process again.");
        error.statusCode = 400;
        throw error;
    }

    const booking = new Booking({
        customerId: bookingIntent.customerId,
        hostId: bookingIntent.hostId,
        listingId: bookingIntent.listingId,
        startDate: bookingIntent.startDate,
        endDate: bookingIntent.endDate,
        totalPrice: bookingIntent.totalPrice,
        finalTotalPrice: bookingIntent.totalPrice,
        bookingStatus: bookingIntent.paymentType === "full" ? "approved" : "pending",
        paymentStatus: bookingIntent.paymentType === "cash" ? "unpaid" :
            bookingIntent.paymentType === "deposit" ? "partially_paid" : "paid",
        paymentMethod: bookingIntent.paymentMethod,
        paymentType: bookingIntent.paymentType,
        depositPercentage: bookingIntent.depositPercentage,
        depositAmount: bookingIntent.depositAmount,
        remainingAmount: bookingIntent.remainingAmount,
        paidAmount: bookingIntent.paymentType === "cash" ? 0 : bookingIntent.paymentAmount,
        transactionId: transactionId || null,
        paidAt: bookingIntent.paymentType !== "cash" ? new Date() : null,
        approvedAt: bookingIntent.paymentType === "full" ? new Date() : null,
        paymentHistory: bookingIntent.paymentType !== "cash" ? [{
            amount: bookingIntent.paymentAmount,
            method: bookingIntent.paymentMethod,
            status: "paid",
            transactionId: transactionId,
            type: bookingIntent.paymentType,
            paidAt: new Date(),
            notes: bookingIntent.paymentType === "full"
                ? "Full payment via VNPay"
                : `Deposit payment (${bookingIntent.depositPercentage}%) via VNPay`,
        }] : [],
    });

    await booking.save();
    await bookingIntent.confirmPayment(transactionId, booking._id);

    return { booking, bookingIntent };
};

const getActiveUserIntents = async (userId) => {
    return await BookingIntent.find({
        customerId: userId,
        status: "locked",
        expiresAt: {$gt: new Date()},
    })
    .populate("listingId", "title listingPhotoPaths price")
    .sort({createdAt: -1});
};

module.exports = {
  releaseExpiredLocks,
  createBookingIntent,
  checkAvailability,
  getByTempOrderId,
  getById,
  cancelIntent,
  confirmIntent,
  getActiveUserIntents
};
